var/spack/repos/builtin/packages/flann/package.py

```diff
@@ -0,0 +1,134 @@
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+
+from spack import *
+
+
+class Flann(CMakePackage):
+    """
+    FLANN is a library for performing fast approximate nearest neighbor
+    searches in high dimensional spaces. It contains a collection of
+    algorithms we found to work best for nearest neighbor search and a system
+    for automatically choosing the best algorithm and optimum parameters
+    depending on the dataset.
+
+    FLANN is written in C++ and contains bindings for the following languages:
+    C, MATLAB and Python.
+
+    NOTE: The author of the package was careful to support HDF5 without
+    parallel support, so the dependencies do not require this, but you will
+    likely want to make sure that you are giving hdf5+mpi as the dependency,
+    this is the spack default so it should be fine.
+    """
+
+    homepage = "http://www.cs.ubc.ca/research/flann/"
+    url      = "https://github.com/mariusmuja/flann/archive/1.9.1.tar.gz"
+
+    version('1.9.1', '73adef1c7bf8e8b978987e7860926ea6')
+    version('1.8.5', '02a81640b1e9c11796a0413976dc11f5')
+    version('1.8.4', '774b74580e3cbc5b0d45c6ec345a64ae')
+    version('1.8.1', '1f51500e172f5e11fbda05f033858eb6')
+    version('1.8.0', '473150f592c2997e32d5ce31fd3c19a2')
+
+    # Options available in the CMakeLists.txt
+    # Language bindings
+    variant("python",   default=False,
+            description="Build the Python bindings. "
+                        "Module: pyflann. "
+                        "Python2 verified. "
+                        "Python3 claimed, not attainable.")
+    variant("matlab",   default=False, description="Build the Matlab bindings.")
+    # default to true for C because it's a C++ library, nothing extra needed
+    variant("c",        default=True,  description="Build the C bindings.")
+
+    # Must build C bindings for Python / Matlab
+    conflicts("+python", when="~c")
+    conflicts("+matlab", when="~c")
+
+    # Additional options
+    variant("cuda",     default=False, description="Build the CUDA library.")
+    variant("examples", default=False, description="Build the examples.")
+    variant("doc",      default=False, description="Build the documentation.")
+    variant("openmp",   default=True,  description="Use OpenMP multi-threading.")
+    # mpi and hdf5 are the bread and butter of this library, use 'em
+    variant("mpi",      default=True,  description="Use MPI.")
+    variant("hdf5",     default=True,  description="Enable HDF5 support.")
+
+    # Dependencies
+    extends("python",      when="+python")
+    depends_on("py-numpy", when="+python", type=("build", "run"))
+    depends_on("matlab",   when="+matlab")
+    depends_on("cuda",     when="+cuda")
+    depends_on("mpi",      when="+mpi")
+    depends_on("hdf5",     when="+hdf5")
```

Comment:
If `flann` defaults to `+hdf5+mpi` and `hdf5` defaults to `+mpi`, then it will do what you want. Of course, users can override all of these defaults in their `packages.yaml`. But if there is no hard requirement, than I agree, keep it as flexible as possible.